modules/core/src/main/java/org/apache/ignite/internal/processors/cluster/GridClusterStateProcessor.java

```diff
@@ -176,6 +179,12 @@
     /** */
     private final ConcurrentMap<UUID, GridFutureAdapter<Void>> transitionFuts = new ConcurrentHashMap<>();
 
+    /**
+     * This latch is used to ensure that on client nodes the {@link ChangeGlobalStateMessage} is fully processed
+     * in the {@link GridCachePartitionExchangeManager} before the {@link ChangeGlobalStateFinishMessage} is processed.
+     */
+    private volatile CountDownLatch changeStateInProgress;
```

Comment:
I reviewed how StateChange messages are created and sent, it seems possible to get two consequent StateChange messages before receiving StateChangeFinished for the first one.

So single latch isn't enough here, we should use ConcurrentHashMap and map latches by request IDs available both in StateChange and StateChangeFinished messages.

Reply:
Done